TF_Service_and_Tooling_Unity/Assets/TouchFree/ServiceUI/Scripts/DiagnosticAPI.cs

```diff
@@ -0,0 +1,227 @@
+            };
+        }
+
+        Debug.Log("DiagnosticAPI connecting... ");
+        try
+        {
+            webSocket.Connect();
+        }
+        catch (System.Exception ex)
+        { 
+            Debug.Log("DiagnosticAPI connection exception... " + "\n" + ex.ToString() );
+            status = Status.Expired;
+        }
+    }
+
+    private void onMessage (object sender, MessageEventArgs e)
+    {
+        if (e.IsText)
+        {
+            newMessages.Enqueue(e.Data);
+        }
+    }
+
+    void HandleMessage(string _message)
+    {
+        if (_message.Contains("GetImageMask"))
+        {
+            try
+            {
+                GetImageMaskResponse maskResponse = JsonUtility.FromJson<GetImageMaskResponse>(_message);
+                OnGetMaskingResponse?.Invoke(maskResponse.value);
+            }
+            catch
+            {
+                Debug.Log("DiagnosticAPI - Could not parse GetImageMask data: " + _message);
+            }
+        }
+        else if(_message.Contains("GetDevices"))
+        {
+            try
+            {
+                GetDevicesResponse devicesResponse = JsonUtility.FromJson<GetDevicesResponse>(_message);
+                connectedDeviceID = devicesResponse.value[0].id;
+                Request("GetImageMask:" + connectedDeviceID);
+            }
+            catch
+            {
+                Debug.Log("DiagnosticAPI - Could not parse GetDevices data: " + _message);
+            }
+        }
+        else if(_message.Contains("Version"))
+        {
+            try
+            {
+                string version = _message.Split('=')[1]; // Split the string (version=X.X.X) and get the 2nd half (X.X.X)
+                // GetVersionResponse versionResponse = JsonUtility.FromJson<GetVersionResponse>(_message); TODO: Change to this for next API version?
+                HandleDiagnosticAPIVersion(version);
+            }
+            catch
+            {
+                Debug.Log("DiagnosticAPI - Could not parse Version data: " + _message);
+            }
+        }
+    }
+
+    public void HandleDiagnosticAPIVersion(string _version)
+    {
+        Version curVersion = new Version(_version);
+        Version minVersion = new Version(minimumMaskingAPIVerison);
+
+        if (curVersion.CompareTo(minVersion) >= 0)
+        {
+            // Version allows masking
+            maskingAllowed = true;
+        }
+        else
+        {
+            // Version does not allow masking
+            maskingAllowed = false;
+        }
+
+        OnMaskingVersionCheck?.Invoke(maskingAllowed);
+    }
+
+    public void Request(string key)
+    {
+        if (status == Status.Connected)
+        {
+            webSocket.Send(key);
+        }
+        else
+        {
+            Connect();
+        }
+    }
+
+    void IDisposable.Dispose ()
+    {
+        status = Status.Expired;
+        webSocket.Close();
+    }
+
+    [Serializable]
+    public struct ImageMaskData
```

Comment:
I'm used to keeping files to a single-class so my first impulse would be to shift these structs out to either an existing type file or a new one, but I noticed that these structs are all declared _inside the class_, does that have any benefits? Does it actually scope them to being only available to the class? If so I'm wondering if we could drop them up at the top above the constructor/member declarations? My general preference is to have the accessories at the top and then get into the meat of the class at the bottom, though if they're _inside_ the type that doesn't seem to work as well. 

* If there's benefit to them being declared _within_ the class I think this is fine
* if there _isn't_ I'd prefer to see them shifted outside of the class and at the top
* if they're referenced outside of this class, that's when I'd prefer them to be out in a types-style file, rather than having to dig to the bottom of this class to find a type to understand it when coming from another class entirely.

Reply:
These structs are specifically utility for the DiagnosticAPI. I believe so long as we keep their use strictly within the DiagnosticAPI class, it makes sense?

Perhaps I should convert them from being public to ensure they are not used externally.

In doing so I should convert the OnGetMaskingResponse event that is triggered and used via CameraPreviewScreen.cs away from using one of the internal structs